### PR TITLE
Handle SES bounce and complaint notifications

### PR DESCRIPTION
**Pull request by vbrown608**
Handles AWS Simple Email Service (SES) bounces and complaints submitted to a webhook via AWS Simple Notification Service (SNS).

Rather than verifying the signature ([something like this](https://github.com/robbiet480/go.sns/blob/master/main.go#L45-L91)) I'm checking for an API key set on SNS notification. That's how we're handling things in the Action Center and I think it's adequate for email bounce/complaint handling but let me know if you'd rather do the full verification.

Separately, we'll need to configure SNS to send notifications with the API key to this endpoint and set that key in the production env.

### CODE REVIEW

File: db/scripts/init_tables.sql [hunk 1]
```diff
@@ -23,3 +23,12 @@ CREATE TABLE IF NOT EXISTS domains
...
 );
+
+
+CREATE TABLE IF NOT EXISTS blacklisted_emails
+(
+    id          SERIAL PRIMARY KEY,
+    email       TEXT NOT NULL,
+    reason      TEXT NOT NULL,
+    timestamp   TIMESTAMP
+);
```
Comment (sydneyli) [inline db/scripts/init_tables.sql]: This reminds me, we should probably talk about how we want to do database migrations. Should all of our scripts just be idempotent?

Comment (vbrown608) [reply db/scripts/init_tables.sql]: It looks like there are a couple libraries for managing migrations but at the moment the schema seems simple enough that we could continue this way until we need one, yeah.

Comment (sydneyli) [reply db/scripts/init_tables.sql]: SG. Just to document our offline convo, we could:
 - Version our database and build tooling to automatically select which migrations are needed to upgrade it.
 - Do the thing we're doing now, which is to write idempotent scripts that run every time the container restarts.

And we decided to continue going with the latter until we have a use case that is complex enough to need the former.

File: email.go [hunk 2]
```diff
@@ -98,9 +105,115 @@ func (c emailConfig) SendValidation(domainInfo *db.DomainData, token string) err
...
+	return nil
+}
+
+// handleSESNotification handles AWS SES bounces and complaints submitted to a webhook
+// via AWS SNS (Simple Notification Service).
+// The SNS webwook is configured to include a secret API key stored in the environment.
+func handleSESNotification(database db.Database) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		keyParam := r.URL.Query()["amazon_authorize_key"]
+		if len(keyParam) == 0 || keyParam[0] != os.Getenv("AMAZON_AUTHORIZE_KEY") {
```
Comment (sydneyli) [inline email.go]: do we 401 all requests to `/sns` if this var isn't set? We might want to do something explicitly; could be as simple/unobtrusive as dropping a message in the `log` just to let the operator know something's up, or returning 5** to let the client know it's not their fault.

Comment (sydneyli) [reply email.go]: also, a follow-up question about this line: is `keyParam` a string? Should it be `keyParam != os.Getenv(...)`?

**Comment by vbrown608** [comment]
Additional note for myself: need to make sure we can complete https://docs.aws.amazon.com/sns/latest/dg/SendMessageToHttp.html#SendMessageToHttp.confirm

File: email.go [hunk 3]
```diff
@@ -98,9 +105,115 @@ func (c emailConfig) SendValidation(domainInfo *db.DomainData, token string) err
...
+			} `json:"complainedRecipients"`
+			Timestamp string `json:"timestamp"`
+		} `json:"complaint"`
+		Bounce struct {
+			BouncedRecipients []struct {
+				EmailAddress string `json:"emailAddress"`
+			} `json:"bouncedRecipients"`
+			Timestamp string `json:"timestamp"`
+		} `json:"bounce"`
+	}
```
[the remark this answers is not in the document]
Comment (sydneyli) [reply email.go]: fun with go structs  :building_construction: 
https://attilaolah.eu/2014/09/10/json-and-struct-composition-in-go/
although i don't really think we should use any of these here. just something i read before i gave up and decided it wasn't worth it!

[the remark this answers is not in the document]
Comment (vbrown608) [reply email.go]: Cool, played around with that a bit [here](https://github.com/EFForg/starttls-backend/commit/a53723dbc13d7d28352e1fd0ed3f61bad94c1479). Let me know if you think it's an improvement.

(same as what I sent you on mattermost, just capturing it here)

[the remark this answers is not in the document]
Comment (sydneyli) [reply email.go]: :+1: Just a comment before the weird section and it should be good to go!

**Comment by sydneyli** [comment]
> Additional note for myself: need to make sure we can complete https://docs.aws.amazon.com/sns/latest/dg/SendMessageToHttp.html#SendMessageToHttp.confirm

Should we open another issue for this?

**Comment by vbrown608** [comment]
I think we can resolve it pretty straightforwardly by logging the full request to Sentry, which could be useful anyway. Then we can just take a look at the body and visit the link within.

**Comment by vbrown608** [comment]
This is ready for re-review :)

File: email_test.go [hunk 4]
```diff
@@ -39,11 +41,100 @@ func TestRequireEnvConfig(t *testing.T) {
...
+		t.Errorf("SES notification without key should return 200, got %d", resp.StatusCode)
+	}
+
+	blacklisted, err := api.Database.IsBlacklistedEmail("complaint@simulator.amazonses.com")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !blacklisted {
+		t.Error("failed to blacklist email")
+	}
```
Comment (sydneyli) [inline email_test.go]: could we add another sanity check to ensure `sendEmail` fails/short-circuits in the correct way?

**Pull request merged**